pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,335 @@
+			cloudProviderReservedIPRanges: map[string]bool{
+				ips[0]: true,
+			},
+			errorExpected: false,
+		},
+		{
+			name:     "1 Pending 1 Used",
+			expected: "192.168.92.16/29",
+			cidr:     "192.168.92.0/27",
+			pendingIPRanges: map[string]bool{
+				ips[0]: true,
+			},
+			cloudProviderReservedIPRanges: map[string]bool{
+				ips[1]: true,
+			},
+			errorExpected: false,
+		},
+		{
+			name:     "2 Pending 1 Used",
+			expected: "192.168.92.24/29",
+			cidr:     "192.168.92.0/27",
+			pendingIPRanges: map[string]bool{
+				ips[0]: true,
+				ips[2]: true,
+			},
+			cloudProviderReservedIPRanges: map[string]bool{
+				ips[1]: true,
+			},
+			errorExpected: false,
+		},
+		{
+			name:     "Pending and used IPs out of CIDR range",
+			expected: "192.168.92.0/29",
+			cidr:     "192.168.92.0/27",
+			pendingIPRanges: map[string]bool{
+				"192.168.33.33/29": true,
+				"192.168.44.44/29": true,
+			},
+			cloudProviderReservedIPRanges: map[string]bool{
+				"192.255.255.0/29":   true,
+				"192.168.255.255/29": true,
+			},
+			errorExpected: false,
+		},
+		{
+			name:     "Unreserved IP Range obtained with carry over to significant bytes",
+			expected: "192.168.2.0/29",
+			cidr:     "192.168.0.0/16",
+			// Using a function for this case as we reserve 32 IP ranges
+			pendingIPRanges: getIPRanges("192.168.0.0/16", 32),
+			// Reserving IP ranges from 192.168.0.0/29 to 192.168.1.248
+			cloudProviderReservedIPRanges: getIPRanges("192.168.1.0/16", 32),
+			errorExpected:                 false,
+		},
+		{
+			name:     "2 Pending 2 Used. Unreserved IPRange unavailable",
+			expected: "",
+			cidr:     "192.168.92.0/27",
+			pendingIPRanges: map[string]bool{
+				ips[0]: true,
+				ips[2]: true,
+			},
+			cloudProviderReservedIPRanges: map[string]bool{
+				ips[1]: true,
+				ips[3]: true,
+			},
+			errorExpected: true,
+		},
+	}
+
+	for _, test := range cases {
+		ipAllocator := initTestIPAllocator()
+		ipAllocator.pendingIPRanges = test.pendingIPRanges
+		ipRange, err := ipAllocator.GetUnreservedIPRange(test.cidr, test.cloudProviderReservedIPRanges)
+		if err != nil && !test.errorExpected {
+			t.Errorf("test %q failed: got error %s, expected %s", test.name, err.Error(), test.expected)
+		} else if err == nil && test.errorExpected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected error", test.name, ipRange)
+		} else if ipRange != test.expected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected %s", test.name, ipRange, test.expected)
+		}
+	}
+}
+
+func getIPRanges(cidr string, ipRangesCount int) map[string]bool {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	ipRangeMask := net.CIDRMask(ipRangeSize, ipV4Bits)
+	i := 0
+	incrementStep := (byte)(math.Exp2(ipV4Bits - ipRangeSize))
+	ipRanges := make(map[string]bool)
+	// Break out of the loop if
+	// 1) We have the required number of IP ranges in the set
+	// 2) IP range overflow occurs and IP increment is not possible
```

Comment:
because these are test inputs, I think if 2 or 3 occurs, that should be a fatal error instead of skipped.